### PR TITLE
feat(qa construct): add optional long context window approach

### PR DESCRIPTION
**Pull request by MichaelWalker-git**
### Problem
Our construct currently defaults to a long context window approach for processing documents when a filename is specified. This method leverages an extensive portion of the document's content as the context for generating AI responses. While effective for comprehensive context provision, it may not always be the most efficient or relevant approach for all use cases.

### Proposed Solution
To provide users with greater flexibility and control over the response generation process, we propose introducing a new parameter within our GraphQL schema: responseGenerationMethod. This parameter will allow users to explicitly select their preferred method of context application—either the existing long context window (longContext) or the Retrieval-Augmented Generation (RAG) method.

### (QA construct): Update Q/A logic #137
https://github.com/awslabs/generative-ai-cdk-constructs/issues/137

### Screenshots:
Screenshot 2024-01-15 at 9 22 58 PM
Screenshot 2024-01-15 at 9 17 53 PM
---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the project license.

### CODE REVIEW

File: resources/gen-ai/aws-rag-appsync-stepfn-opensearch/schema.graphql
Comment (krokoko) [inline resources/gen-ai/aws-rag-appsync-stepfn-opensearch/schema.graphql:1]: seems like this is the incorrect file being modified (should be https://github.com/awslabs/generative-ai-cdk-constructs/blob/main/resources/gen-ai/aws-qa-appsync-opensearch/schema.graphql) ? the graphql schema should be updated only for QA, this feature doesn't impact the ingestion pipeline.

File: lambda/aws-rag-appsync-stepfn-opensearch/embeddings_job/src/lambda.py
Comment (krokoko) [inline lambda/aws-rag-appsync-stepfn-opensearch/embeddings_job/src/lambda.py:1]: Same here, the ingestion pipeline should not be modified, impact is only on QA

File: resources/gen-ai/aws-rag-appsync-stepfn-opensearch/schema.graphql
Comment (MichaelWalker-git) [reply resources/gen-ai/aws-rag-appsync-stepfn-opensearch/schema.graphql:1]: whoops. Fixing now.

File: lambda/aws-rag-appsync-stepfn-opensearch/embeddings_job/src/lambda.py [hunk 1]
```diff
@@ -92,7 +92,7 @@ def handler(event,  context: LambdaContext) -> dict:
```
Comment (MichaelWalker-git) [inline lambda/aws-rag-appsync-stepfn-opensearch/embeddings_job/src/lambda.py:92]: My IDE linter automatically applied some changes. Sorry for the noise.

**Comment by krokoko** [comment]
Test cases discussed:
- ResponseGenerationMethod set to RAG should run the RAG approach (regardless of a filename specified)
- ResponseGenerationMethod set to Long context and no filename specified should send back an error message
- ResponseGenerationMethod set to Long context window and a filename specified should check how many tokens are in the specified document, if that amount is smaller that the max number of tokens supported by the model used, run Long Context window, otherwise run RAG approach

To know which approach was used, this information will be logged by the Lambda function

**Review by scottschreckengaust — APPROVED** [review]

**Review by dineshSajwan — APPROVED** [review]

**Pull request merged**